README.md
# Brane Standard Library
This repository provides a standard collection of packages for a running [Brane](https://github.com/epi-project/brane) instance.


## Repository structure
The individual packages are located under each folder in the repository root under the same name as their Brane identifier. For example, the `cat` package is available in the folder `/cat`. Each folder contains a README.md with more information about that package.

For a brief overview, see the table [below](#overview).

To download a package, you can run the following command on your local machine (we assume that you have the [Brane Command-Line Interface](https://github.com/epi-project/brane/releases/latest) installed):
```bash
brane import epi-project/brane-std ./<package_name>/container.yml
```
where you should replace `<package_name>` with the name of the package you like to install (e.g., `cat`).

> <img src="./assets/img/info.png" alt="info" width="16" style="margin-top: 3px; margin-bottom: -3px;"/> If you are running an unreleased version of Brane, you can add `--init <path/to/branelet>` to use a custom-compiled `branelet` executable, just as you can with `brane build`.


## Overview
This repository currently includes the following packages:
- `cat`: A package for quickly inspecting dataset contents. Used in the [official Wiki](https://wiki.enablingpersonalizedinterventions.nl).
- `copy_result`: A "necessary evil" package that takes an intermediate result and renames it. Necessary if you are writing to the same result as you're reading from (i.e., a loop).
- `data_init`: A package that contains functions for generating intermediate results of various shapes.
- `data_math`: A package that contains functions for doing arithmetic on datasets or intermediate results of a certain shape.
- `hello_world`: A package that contains the simplest function of all. Used in the [official Wiki](https://wiki.enablingpersonalizedinterventions.nl).


## Contributing
If you like to contribute to this repository by adding your own package, please make a pull request where you provide your package in the same structure as provided above. Please use `TEMPLATE.md` to write the README for your package, and include a `CHANGELOG.md` based on [this](https://keepachangelog.com/en/1.0.0/) format so users can see what has changed.

If you have suggestions about the individual packages, feel free to raise issues or create pull requests as well. Any help is appreciated!


## Attribution
The icons used in this repository (<img src="./assets/img/info.png" alt="info" width="16" style="margin-top: 3px; margin-bottom: -3px;"/>, <img src="./assets/img/warning.png" alt="warning" width="16" style="margin-top: 3px; margin-bottom: -3px;"/>) are provided by [Flaticon](https://flaticon.com).


data_init/CHANGELOG.md
# CHANGELOG for the `data_init` package

This file contains a log of the changes made for the various versions of the Data Init package.

Note that the package's versioning scheme uses [semantic versioning](https://semver.org).

For more information about the package itself, see [README.md](/epi-project/brane-std/blob/main/data_init/README.md).


## 1.0.0 - 2023-01-16
### Added
- Initial release.
- The `zeroes()` function.


cat/code.py
#!/usr/bin/env python3
# CODE.py
#   by Lut99
#
# Created:
#   06 Dec 2022, 15:14:47
# Last edited:
#   06 Dec 2022, 15:41:34
# Auto updated?
#   Yes
#
# Description:
#   Implements the code for the `cat` package (see the README.md file).
#

import base64
import json
import os
import sys


##### FUNCTION FUNCTIONS #####
def cat(dataset: str, path: str) -> str:
    """
        Implements the `func cat(dataset, path)` function.
    """

    # Compute the total path
    path = os.path.join(dataset, path) if path != "-" else dataset

    # Attempt to open the file
    try:
        with open(path, "r") as h:
            return h.read()
    except IOError as e:
        print(f"Failed to read file '{path}': {e}", file=sys.stderr)
        exit(1)

def cat_base64(dataset: str, path: str) -> str:
    """
        Implements the `func cat_base64(dataset, path)` function.
    """

    # Compute the total path
    path = os.path.join(dataset, path) if path != "-" else dataset

    # Attempt to open the file
    try:
        with open(path, "rb") as h:
            # Return it encoded
            return base64.b64encode(h.read()).decode("UTF-8")
    except IOError as e:
        print(f"Failed to read file '{path}': {e}", file=sys.stderr)
        exit(1)

def cat_range(dataset: str, path: str, start: int, end: int) -> str:
    """
        Implements the `func cat_range(dataset, path, start, end)` function.
    """

    # Compute the total path
    path = os.path.join(dataset, path) if path != "-" else dataset

    # Attempt to open the file
    try:
        with open(path, "r") as h:
            # Seek the correct position first
            h.seek(start)
            return h.read(1 + end - start)
    except IOError as e:
        print(f"Failed to read file '{path}': {e}", file=sys.stderr)
        exit(1)

def cat_range_base64(dataset: str, path: str, start: int, end: int) -> str:
    """
        Implements the `func cat_range_base64(dataset, path, start, end)` function.
    """

    # Compute the total path
    path = os.path.join(dataset, path) if path != "-" else dataset

    # Attempt to open the file
    try:
        with open(path, "rb") as h:
            # Seek the correct position first
            h.seek(start)
            # Return it encoded
            return base64.b64encode(h.read(1 + end - start)).decode("UTF-8")
    except IOError as e:
        print(f"Failed to read file '{path}': {e}", file=sys.stderr)
        exit(1)





##### ENTRYPOINT #####
if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <cat|cat_base64|cat_range|cat_range_base64>")
        exit(1)
    command = sys.argv[1]

    # Read the input arguments
    dataset = json.loads(os.environ["DATASET"])
    path    = json.loads(os.environ["NESTED_PATH"])

    # Run the function
    if command == "cat":
        # Don't forget to replace newlines with newlines with proper indentation, so YAML understands it's all the same file
        content = cat(dataset, path).replace("\n", "\n  ")
        print(f"output: |\n  {content}")
    elif command == "cat_base64":
        # Don't forget to replace newlines with newlines with proper indentation, so YAML understands it's all the same file
        content = cat_base64(dataset, path).replace("\n", "\n  ")
        print(f"output: |\n  {content}")
    elif command == "cat_range":
        # Also read the start and end now
        start = json.loads(os.environ["START"])
        end   = json.loads(os.environ["END"])

        # Don't forget to replace newlines with newlines with proper indentation, so YAML understands it's all the same file
        content = cat_range(dataset, path, start, end).replace("\n", "\n  ")
        print(f"output: |\n  {content}")
    elif command == "cat_range_base64":
        # Also read the start and end now
        start = int(os.environ["START"])
        end   = int(os.environ["END"])

        # Don't forget to replace newlines with newlines with proper indentation, so YAML understands it's all the same file
        content = cat_range_base64(dataset, path, start, end).replace("\n", "\n  ")
        print(f"output: |\n  {content}")


data_init/zeroes.py
#!/usr/bin/env python3
# ZEROES.py
#   by Lut99
#
# Created:
#   16 Jan 2023, 13:16:26
# Last edited:
#   16 Jan 2023, 13:44:09
# Auto updated?
#   Yes
#
# Description:
#   Implements a generator for an intermediate result of only zeroes (ASCII
#   `0`).
#

import argparse
import sys


##### GENERATION FUNCTIONS #####
def generate_vector(n: int) -> int:
    """
        Generates a zeroes file with the vector layout.

        Specifically, generates a file /result/data with `n` zeroes, delimited
        by spaces. Every zero is simply the ASCII character for `0`.

        # Arguments
        - `n`: The number of ASCII `0` to generate.

        # Returns
        The return code of the operation. `0` means success.
    """

    # Attempt to open the output file
    try:
        with open("/result/data", "w") as h:
            # Write the number of zeroes
            for i in range(n):
                if i > 0: h.write(" ")
                h.write("0")

    except IOError as e:
        print(f"Failed to write to output file '/result/data': {e}", file=sys.stderr)
        return e.errno

    # Done
    return 0





##### ENTRYPOINT #####
def main(n: int, kind: str) -> int:
    """
        Entrypoint to the script.

        # Arguments
        - `n`: The number of zeroes to generate.
        - `kind`: The kind of dataset to generate.

        # Returns
        The exit code for the script. `0` means success.
    """

    # Match on the kind
    if kind == "vector":
        return generate_vector(n)

    # Should never happen
    raise RuntimeError(f"main() saw a non-allowed kind '{kind}' (should have been taken care of by the argument parser)")



# The actual entrypoint
if __name__ == "__main__":
    # Parse the arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("NUMBER", type=int, help="The number of zeroes to generate in the file.")
    parser.add_argument("KIND", choices=["vector"], help="The kind of dataset to generate.")
    args = parser.parse_args()

    # Verify numbers is in-of-bounds
    if args.NUMBER < 0:
        print(f"NUMBER has to be a non-negative integer, not {args.NUMBER}", file=sys.stderr)
        exit(1)

    # Call main
    exit(main(args.NUMBER, args.KIND))


data_math/entrypoint.sh
#!/bin/bash
# ENTRYPOINT.sh
#   by Lut99
#
# Created:
#   16 Jan 2023, 13:13:39
# Last edited:
#   16 Jan 2023, 13:41:34
# Auto updated?
#   Yes
#
# Description:
#   Simple script that handles parsing the input and call the correct Python
#   file.
#


# Read the command
if [[ "$#" -ne 1 ]]; then
    2>&1 echo "Usage: $0 <FUNCTION>"
    exit 1
fi
func="$1"

# Read the kind
kind=$(echo "$KIND" | python3 -c "import json, sys; print(json.load(sys.stdin))")

# Match on the function
if [[ "$func" == "add_const" ]]; then
    # Read the dataset and constant from the environment variables
    data=$(echo "$DATA" | python3 -c "import json, sys; print(json.load(sys.stdin))")
    const=$CONST

    # Call the python script
    python3 ./add_const.py "$data" "$const" "$kind"

else
    2>&1 echo "Unknown function '$func'"
    exit 1
fi


data_math/add_const.py
#!/usr/bin/env python3
# ADD CONST.py
#   by Lut99
#
# Created:
#   16 Jan 2023, 13:41:26
# Last edited:
#   16 Jan 2023, 13:50:18
# Auto updated?
#   Yes
#
# Description:
#   Implements an element-wise addition of a dataset and some constant.
#

import argparse
import os
import sys


##### GENERATION FUNCTIONS #####
def add_vector(data: str, const: float) -> int:
    """
        Generates a zeroes file with the vector layout.

        Specifically, generates a file /result/data with `n` (ASCII) constants,
        delimited by spaces.  Their value is the element-wise addition of the
        input `data` and the given `const`.

        # Arguments
        - `data`: The path to the dataset folder that we are operating on. We assume there is a `data` file within it.
        - `const`: The constant to add to each element.

        # Returns
        The return code of the operation. `0` means success.
    """

    # Read the input
    try:
        with open(os.path.join(data, "data"), "r") as source:
            vec = source.read()
    except IOError as e:
        print(f"Failed to read from input file '{os.path.join(data, 'data')}': {e}", file=sys.stderr)
        return e.errno

    # Split it and attempt to parse as real values
    vec = [ float(v) for v in vec.split() ]

    # Add the constant
    vec = [ v + const for v in vec ]

    # Write back, trying to write as integers whenever possible
    try:
        # Attempt to open the output file
        with open("/result/data", "w") as target:
            for i in range(len(vec)):
                if i > 0: target.write(" ")
                if float(int(vec[i])) == vec[i]:
                    target.write(f"{int(vec[i])}")
                else:
                    target.write(f"{vec[i]}")
    except IOError as e:
        print(f"Failed to write to output file '/result/data': {e}", file=sys.stderr)
        return e.errno

    # Done
    return 0





##### ENTRYPOINT #####
def main(data: str, const: float, kind: str) -> int:
    """
        Entrypoint to the script.

        # Arguments
        - `data`: The path to the dataset folder that we are operating on.
        - `const`: The constant to add to each element.
        - `kind`: The kind of dataset to generate.

        # Returns
        The exit code for the script. `0` means success.
    """

    # Match on the kind
    if kind == "vector":
        return add_vector(data, const)

    # Should never happen
    raise RuntimeError(f"main() saw a non-allowed kind '{kind}' (should have been taken care of by the argument parser)")



# The actual entrypoint
if __name__ == "__main__":
    # Parse the arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("DATA", type=str, help="The path to the input data directory.")
    parser.add_argument("CONST", type=float, help="The constant to add to each element.")
    parser.add_argument("KIND", choices=["vector"], help="The kind of dataset to generate.")
    args = parser.parse_args()

    # Call main
    exit(main(args.DATA, args.CONST, args.KIND))
